### ISSUE TITLE
Feature request : proper internal resources management

### ISSUE THREAD

**Issue by MathiasPaulin** [STORM-IRIT/Radium-Engine#337]
## Objective of the Request feature.

The Radium Engine and the plugins or applications built upon Radium, require several file resources to be available at runtime.
Finding a way to structure and to abstract these resources, to allow easy development and efficient deploy is mandatory.

This request feature is a first proposal for robust and efficient resource management and structuring into Radium, its plugins and applications.

## What are resources ?

When we speak about resources, we here consider the following resources. The resource list might later be extended in the discussion.
- Shaders : this is the main resources the Radium depends on when using the engine. Shaders are text files split into several components allowing code re-use and dynamic specialization of their content.
- Data file : Even if these are not in the engine, some plugins or applications need external (precomputed) data to function. A data file is then an octet stream, which structure is usage dependent that will be accessed at runtime

## Actual state :

Shaders are for the moment accessed at runtime explicitly from the OS filesystem.

All Engine, gui and application related shaders are located into the `Shaders` directory into the Radium sources hierarchy that is linked or copy into the `bin` directory at compile time to be accessed at runtime through the standard input/output functionalities of C++.

Even if this is efficient for development, this cause several difficulties
- Code maintenance : as everything is melt in this directory, it is difficult to maintain all the shader code as we do not know which part of Radium need which shader (Engine, Gui, Applications, Plugins ... ). More, as these shaders were programmed independently by several developers, there is not so much code re-using and when the Engine require an evolution of the shader code one have to go through a constantly increasing number of files to fix the code.
- Application deployment : When we want to deploy an application (eg, a technological demo), we have to ensure that we could reproduce the same directory structure from the Radium source hierarchy into the target installation or executable directory. This is not a good idea as we then must expose a lot of files unrelated to the demo (Radium internal shaders)
- System independence : accessing directly the OS file system prevent Radium to be deployed on systems with specific file systems (e.g IOS) or without access to a filesystem. More, each OS put different constraints on the way an application must be deployed (e.g. MacOs .app structure), and how it might access to the filesystem.

## File resources management proposal

This proposal aims at defining the structure and access method to Radium related resources both for an easy development and for efficient deployment. It the propose in source structuring and how to transform this structure to a runtime collection integrated in the deployed binaries.

### In source structure :
Instead of the (contains everything) directory `Shaders`, we will define the following hierarchy in the root Radium source directory :

```
Radium-Engine
└───src
│   └───...
└───Resources
│   └───Engine
│   │   └───Materials
│   │   │   └───BlinnPhong
│   │   │   │   │   BlinnPhong.frag.glsl
│   │   │   │   │   BlinnPhong.vert.glsl
│   │   │   │   │   ...
│   │   │   └───OrenNayar
│   │   │   │   │   ...
│   │   │   ...
│   │   └───Lights
│   │   │   │   ...
│   │   └───Transforms
│   │   │   │   ...
│   │   └───Data
│   │   │   │   ...
│   └───GuiBase
│   │   │   ...
│   └───Core
│   │   │   ...
│   │   ...
...
```
Each application or plugin could the, in its source hierarchy define the same structure :

```
AppOrPlugin
└───src
│   └───...
└───Resources
│   └───AppOrPluginName
│   │   │   ...
...
```
When the binary is generated, each entry of the `Resources` folder, from the Engine, the application and the plugins, are linked (or copied on system that do not support links) into the `bin/Resources` directory to reproduce the same hierarchy.
Then, each entry in the `bin/Resources` directory correspond to the name of the owner of the resources one can access below this entry.

### Runtime (deployed) access :
In the `Core` library, add a `ResourcesManager` class that will give an abstract access the `Resources` directory structure.
- When compiled in a `RADIUM_DEVELOPMENT` mode (by, e.g., defining this symbol at compile time), the access will be a direct access to the filesystem through the fully qualified name of the file (e.g. `Engine/Material/BlinnPhong/BlinnPhong.frag.glsl`).
- When compiled in a `RADIUM_DEPLOY` mode, the access will be done through a map that associate the fully qualified name of the resource (e.g. `Engine/Material/BlinnPhong/BlinnPhong.frag.glsl`) to the content of the file. This map will be static (and so, the resources will be in read-only access) and generated at compile-time. For this, at compile-time, a script will take as input a file containing, for the current component being compiled, the fully qualified name of all the required resources.
This script will then generate a C++ source file that defines a static `std::map<std::string, const char*>` containing the component (e.g. `Core`, `Engine`, `MyPlugin`, ...) resources. The first element of the pair is the name of the resource, the second is a `const char*` corresponding to the content of the file associated to the resource _(note that this will not allow to have binary resources, but we could define a similar approcahc for this)_. In this C++ file, there will be also a function that register the resources to the `ResourcesManager`. This function must then be called when initializing the component so that resources are made available for the current generated binary.

** Notes **
- The proposed `ResourcesManager` and the way it is parameterized at compile time is quite similar to the `.rc` resource file of Qt. But we need to keep independence from Qt as not all applications are qt based. Some are command line.
- All access to files that correspond to resources **must** then be done through the `ResourcesManager`
- Binary resources (e.g. images, binary data files, ...) does not fit in the current proposal but are unused for the moment and we could first validate the process using text-based resources and extend the `ResourcesManager` later to manage binary resources.

**Comment by MathiasPaulin** [comment]
Reopened as not yet solved. 1st step towards resources management was added in PR #448 but hints expressed in this issue are still valid.

**Comment by MathiasPaulin** [comment]
There will be some advances on Resources management on the PR #462. As the buildchain and installation procedure will be refactored, the Resources management has also changed. As soon as the PR #462 will be merged, the current proposal will be transformed in a documentation in the Radium distribution.

**Comment by nmellado** [comment]
Still waiting for PR #476 to be merged.

**Comment by dlyr** [comment]
Fixed by #741 
If not please open a new issue explaining what is the issue wrt current master.
